client/app.js

```diff
@@ -31,46 +31,33 @@ export function bootstrap(config) {
     authorizeUrl: authzUrl,
   });
 
-  const getScopes = (scopeString) => {
-    const xs = decodeURIComponent(scopeString).split('+');
-
-    return xs.filter(x => ['openid', 'profile'].indexOf(x) === -1);
+  const renderView = (accessTokenObj, userInfo) => {
+    // render main view
+    const containerEl = document.querySelector(config.container);
+    const app = Elm.Main.embed(containerEl, {
+      accessTokenResp: accessTokenObj,
+      userInfo: userInfo,
+    });
+    // Elm -> JS
+    app.ports.loginRedirect.subscribe(() => {
+      loginRedirect(auth);
+    });
   };
 
-  // read token from URL
-  let hashObj = null;
-
-  if (window.location.hash) {
-    const hashes = window.location.hash.substr(1).split('&');
-    hashObj = hashes.reduce((init, x) => {
-      const xs = x.split('=');
-      let key = xs[0];
-      const val = xs[1];
-      return Object.assign({}, init, {[key]: val});
-    }, {});
-  }
-
-  const tokenResp = hashObj ? {
-    accessToken: hashObj['access_token'],
-    idToken: hashObj['id_token'],
-    scope: getScopes(hashObj['scope']),
-  } : null;
-
-  console.log(hashObj);
+  auth.token.parseFromUrl()
+    .then(function(tokens) {
+      const accessTokenObj = tokens[0];
 
-  // render main view
-  const containerEl = document.querySelector(config.container);
-  const app = Elm.Main.embed(containerEl, {
-    config: {
-      userInfoUrl
-    },
-    tokenResp: tokenResp,
-  });
-
-  // Elm -> JS
-  app.ports.loginRedirect.subscribe(() => {
-    loginRedirect(auth);
-  });
+      return auth.token.getUserInfo(accessTokenObj)
```

Comment:
get userinfo seems not necessary since it's assembled into the idtoken.

Reply:
- import infor `name` and `preferred_username` are in id token as well.
- not big deal and keep it as for now.